lib/pbsnodes_client.rb

```diff
@@ -0,0 +1,80 @@
+# Utility class for getting numerical data from pbsnodes
+#
+# @author Kinan AlAttar
+require 'open3'
+
+class PBSNodesClient
+
+    attr_accessor :gpus_used, :gpus_available
+
+    # Set the object to the server.
+    #
+    # @param [OodAppkit::Cluster]
+    #
+    # @return [PBSNodesClient] self
+    def initialize(cluster)
+     @server = cluster.custom_config[:pbs]
+     @cluster_id = cluster.id
+     @cluster_title = cluster.metadata.title || cluster.id.titleize
+     self
+    end
+
+    def setup
+      self.gpus_available = calc_available_gpus
+      self.gpus_used = calc_gpus_used
+      self
+    rescue => e
+        PBSNodesClientNotAvailable.new(cluster_id, cluster_title, e)
+    end
+
+    def pbs_bin
+      Pathname.new(@server['bin'].to_s)
+    end
+
+    # Defines a method for counting values of gpu_state
+    #
+    # @param[first_cmd_arg]
+    def gpu_states(first_cmd_arg)
+        cmd = pbs_bin.join("pbsnodes").to_s
+        args = ["-s", first_cmd_arg , ":gpu"]
+        o, e, s = Open3.capture3(cmd, *args)
+        s.success? ? o.scan(/gpu_state=(\w+)/) : raise(CommandFailed, e)
+    rescue Errno::ENOENT => e
+       raise InvalidCommand, e.message
+    end
+
+    # Defines a method for counting values of documented_threads
+    #
+    # @param[first_cmd_arg]
+    def dedicated_threads(first_cmd_arg)
+        cmd = pbs_bin.join("pbsnodes").to_s
+        args = ["-s", first_cmd_arg , ":gpu"]
+        o, e, s = Open3.capture3(cmd, *args)
+        s.success? ? o.scan(/dedicated_threads = ../) : raise(CommandFailed, e)
+    rescue Errno::ENOENT => e
+      raise InvalidCommand, e.message
+    end
+
+
+    # Count available number of GPUs on a cluster
+    def calc_available_gpus
+       gpu_states(@cluster_title + "-batch.ten.osc.edu").size
+    end
+
+    # Count number of GPUs used
+    def calc_gpus_used
+      if @cluster_title != "Pitzer"
+          gpu_states(@cluster_title + "-batch.ten.osc.edu").select {|state| state.include?("Unallocated")}.size
+      else
+          dedicated_threads(@cluster_title + "-batch.ten.osc.edu").select {|thread| thread.include?("40")}.size +

```

Comment:
what is the rationale of:

```
dedicated_threads(@cluster_title + "-batch.ten.osc.edu").select {|thread| thread.include?("40")}.size
```

for Pitzer?

Reply:
Yes, exactly.

Reply:
No I mean please explain the rationale. What does that line mean and why does it exist?

Reply:
The line is used to see how many jobs use all the cores on a specific GPU node, since each node has two GPUs on the Pitzer cluster. So, if a job is using one of the GPUs on a node along with all of the cores (40) on it, then the other GPU cannot be allocated.

Reply:
So the method says "calc_gpus_used" and comment "Count number of GPUs used".

But both appear to be counting the number of gpus with states "Unallocated" which would be the opposite - gpus not used. And in the Pitzer example, if you count the number of gpu nodes that are fully allocated, that would mean the number of gpus that are allocated is `num_gpu_nodes_allocated*2`.

So it seems like the logic here is a little off...

Reply:
I completely overlooked this as I was initially thinking of implementing the class and its methods differently. I will fix this. thanks.